[이채빈] Week5

caseV12 opened the pull request:
## 요구사항

### 기본

- [x] [기본]https://bootcamp-api.codeit.kr/docs 에 명세된 “/api/sign-in”으로 { “email”: “test@codeit.com”, “password”: “sprint101” } POST 요청해서 성공 응답을 받고 “/folder”로 이동하나요?
- [x] [기본]이메일 input에서 “test@codeit.com”으로 “/api/check-email” 이메일 중복 확인 POST 요청하면 이메일 중복 에러를 확인할 수 있나요?
- [x] [기본]유효한 회원가입 형식의 경우 “/api/sign-up” POST 요청하고 성공 응답을 받으면 “/folder”로 이동하나요?

### 심화

- [x] [심화]로그인/회원가입시 성공 응답으로 받은 accessToken을 로컬 스토리지에 저장했나요?
- [x] [심화]로그인/회원가입 페이지에 접근시 로컬 스토리지에 accessToken이 있는 경우 “/folder” 페이지로 이동하나요?

## 주요 변경사항

- 로그인, 회원가입 페이지에서 서버 통신 추가
- accessToken에 따른 페이지 전환 추가

## 스크린샷


## 멘토에게

- 그동안 감사했습니다! 프로그래밍 뿐만 아니라 여러 방면에서 배워가는 부분이 많았습니다 감사합니다! 
-
-

caseV12 committed:
feat: signin.js에 input태그 이벤트리스너 추가

caseV12 committed:
feat: signin 페이지 input 태그 pattern, class 추가

caseV12 committed:
style: sign.css에 input 에러 관련 디자인 추가

caseV12 committed:
feat: signup.js에 validPassCheck 이벤트핸들러 추가

caseV12 committed:
feat: "로그인폼 submit 핸들러 추가"

caseV12 committed:
feat: 회원가입폼 submit 핸들러 추가

caseV12 committed:
feat: 비밀번호 숨김 버튼 토글 기능 추가

caseV12 committed:
refactor: 스크립트 파일 모듈화 및 리팩터링

caseV12 committed:
Refactor: togglePasswordHidden 함수의 마크업 구조로부터 독립성을 향상

caseV12 committed:
Refactor: signForm 유효성 검증 로직 개선

커밋 전 매 이벤트마다 불필요하게 노드를 지우고 생성했던 로직을 없앰.
노드를 제거하는 로직은 validate 모듈의 early return 아래로 내림.
renderErrorMSGNode 함수에서는 기존 노드와 에러메시지 확인 후 이에 맞게 작동하도록 함.
이를 통해 불필요한 동작 개선.

함수명이 이벤트와 연결되도록 개선.

caseV12 committed:
Refactor: 모듈화 전 스크립트 삭제

caseV12 committed:
Feat: 로그인 폼 제출 시 api/sign-in 통신 함수 postSignIn 추가

caseV12 committed:
feat: 회원가입 폼 제출시 api/sign-up api/check-email 통신 추가

api/check-email에 post 요청을 보내고 status 정보가 담긴 객체를 리턴하는 postCheckEmail 함수 추가
postCheckEmail의 리턴에 따라 api/signup에 post 요청을 보내고 json.data를 받는 postSignUp 함수 추가

caseV12 committed:
feat: accessToken에 따라 페이지 전환 기능 추가

Tanney-102 commented on scripts/modules/handleErrorClass.js:
```diff
@@ -0,0 +1,31 @@
+
+function renderErrorMSGNode(target, msg){
+    target.classList.add(CLASSNAME_ERROR)
+    const parentLabel = target.closest(".signForm_label");
```
오 closest도 나름 좋은 접근이네요.
그런데 parentLabel은 target에 관련없이 항상  .signForm_label을 클래스네임으로 갖나요?

Tanney-102 commented on scripts/modules/redirector.js:
```diff
@@ -0,0 +1,8 @@
+function passAuth(){
+    const existAccessToken = localStorage.getItem("accessToken")
+    if(existAccessToken){
+        location.assign("./folder.html")
+    }
+}
```
passAuth라는 이름만 보고는 /folder 로 redirect 한다는 걸 유추하기 어렵습니다.
함수 이름을 바꾸거나 인자를 추가하는 방식으로 더 명시적인 함수를 작성할 수 있을 것 같아요.
간단한 예시 남겨드립니다.

```js
function redirect(to, predicate) {
  if (predicate) {
    location.assign(to)
  }
}

function redirectWithAuth(to) {
  const existAccessToken = localStorage.getItem("accessToken")
  redirect(to, existAccessToken)
}
```

Tanney-102 commented on scripts/modules/validateSignForm.js:
```diff
@@ -0,0 +1,44 @@
+import { renderErrorMSGNode, removeErrorMSGNode } from "./handleErrorClass.js";
+
+const passwordInputNode = document.querySelectorAll(".signForm_input.password")[0]
```
하나로 특정할 수 있는 classname을 추가하고 querySelector를 쓰는 편이 좋겠습니다.
index order에 의존하는 방법은 피해주세요

Tanney-102 commented on scripts/modules/validateSignForm.js:
```diff
@@ -0,0 +1,44 @@
+
+const passwordInputNode = document.querySelectorAll(".signForm_input.password")[0]
+
+function validateEmailOnFocusout(e){
```
함수이름 굳이네요 👍

Tanney-102 commented on scripts/signin.js:
```diff
@@ -0,0 +1,53 @@
+    const email = e.target.email.value
+    const password = e.target.password.value
+    const payload = {email, password}
+    const postSignIn = async (payload) => {
+        try {
+            const response = await fetch("https://bootcamp-api.codeit.kr/api/sign-in",{
+                method: "POST",
+                headers: { 'Content-Type': 'application/json'},
+                body: JSON.stringify(payload),
+            });
+          
+            if (!response.ok) {
+                throw new Error('Network response was not ok');
+            }
+            const json = await response.json()
+            const {accessToken} = json.data
+            localStorage.setItem("accessToken",JSON.stringify(accessToken)); 
+            location.assign("./folder.html")
+            return json.data
+        } catch (error) {
+            console.error('There was a problem with the fetch operation:', error);
+        };
```
밖에 선언해도 될 것 같아요

Tanney-102 commented on scripts/signup.js:
```diff
@@ -0,0 +1,73 @@
+        try {
+            const responseOfCheckEmail = await postCheckEmail(email_payload)
+            if(responseOfCheckEmail.status === 409){
+                renderErrorMSGNode(emailInputNode, "이미 사용중인 이메일입니다.")
```
에러 핸들링을 사용처에서 하는 게 좋을 듯 합니다. 지금은 이 함수가 하는 일이 너무 많아요. 함수의 관심사는 가급적 하나로 제한해주세요

Tanney-102 reviewed (commented):
채빈님!
리뷰 늦어서 죄송합니다ㅜㅜ
함수와 모듈을 깔끔하게 잘 나눠주셨네요 👍 
몇가지 코멘트 남겼으니 확인 부탁드려요!

아 그리고 컨플릭도 해결 부탁드립니다!

caseV12 committed:
conflict 해결

caseV12 committed:
Refactor: redirector 모듈 함수명 변경

caseV12 committed:
Refactor: sign 페이지에 class 추가로 마크업 의존성 감소

signin, signup 페이지의 요소에 접근할 때
공통된 class명과 index order를 사용하는 방법에서
index order 없이 고유한 class명을 사용하여 접근하는 것으로 변경

caseV12 committed:
signin/up.js의 이벤트핸들러 내부 불필요한 함수 선언 부분 외부로 분리
